### ISSUE TITLE
RESTSerializer throws on pushing record with relationship

### ISSUE THREAD

**Issue by jelhan** [samselikoff/ember-cli-mirage#972]
Following docs ember-cli-mirages `RestSerializer` should be used "to fake backends that match Ember Data’s RestSerializer expected response format". But if using it together with default ember-data RESTAdapter and RESTSerializer it throws on pushing records with relationship:

> Mirage: Your handler for the url /comments threw an error: newModel.isNew is not a function

I've set up an ember twiddle demonstrating the issue: https://ember-twiddle.com/328e1448a6a50e8372f0709b4a4be3f6

I tried to debug the issue but I'm not quite sure whats the expected behavior. I figured out that ember-cli-mirage throws cause it expects an model instance [here](https://github.com/samselikoff/ember-cli-mirage/blob/master/addon/orm/associations/belongs-to.js#L106-L118) but it's called with an id.

Also ember-cli-mirage RestSerializer does not seem to follow ember-data RESTSerializer as stated in docs. ember-cli-mirage RestSerializer seem to expect an "Ids" suffix[1] which ember-data RESTSerializer does not use by default[2]. But overriding `keysForRelationshipIds` method does not fix the issue. To test that one just replace `mirage/serializers/applications.js` in twiddle with this one:
```
import { RestSerializer } from 'ember-cli-mirage';
import { camelize, singularize } from 'ember-cli-mirage/utils/inflector';

export default RestSerializer.extend({
  keyForRelationshipIds(type) {
    return camelize(singularize(type));
  }
});
```

Digging deeper into it, `normalize` method of `ActiveModelSerializer` which is also used by `RestSerializer` doesn't seem to return a valid JSONApiDocument. In my example it's called with this payload:

```
{
  "comment": {
    "post": "2"
  }
}
```

And returns this object:

```
{
  "data": { 
    "type": "comments",
    "attributes": {
       "post": "2"
    }
  }
}
```

As far as I got JSON Api specs relationships should not be part of `attributes` object. They should be included in `relationships` object as `data` property.

Also it seem to be impossible to override the key used for a belongs-to relationship. It's suffixed by "Id" and there doesn't seem to be any `keyForRelationshipId` method.

[1] [addon/serializers/rest-serializer.js](https://github.com/samselikoff/ember-cli-mirage/blob/master/addon/serializers/rest-serializer.js#L22-L24) and [tests/integration/serializers/rest-model-serializer-test.js](https://github.com/samselikoff/ember-cli-mirage/blob/24f994f4419f1b175c82331b00c3089e3f65c038/tests/integration/serializers/rest-model-serializer-test.js#L50)
[2] [DS.RESTSerliazer normalize docs](http://emberjs.com/api/data/classes/DS.RESTSerializer.html#method_normalize)

**Comment by jelhan** [comment]
I played a little bit around. If I override `normalize` method to return a valid JSON Api document also for relationships it seem to work fine. I just hacked it to test:

```
  normalize(payload) {
    let type = Object.keys(payload)[0];
    let attrs = payload[type];

    let jsonApiPayload = {
      data: {
        type: pluralize(type),
        attributes: {}
      }
    };
    if (attrs.id) {
      jsonApiPayload.data.id = attrs.id;
    }
    Object.keys(attrs).forEach(key => {
      if (key !== 'id') {
        jsonApiPayload.data.attributes[dasherize(key)] = attrs[key];
      }
    });
      
    if (isPresent(jsonApiPayload.data.attributes['post'])) {
      jsonApiPayload.data.relationships = {
        "post": {
          "data": { "type": "posts", "id": jsonApiPayload.data.attributes['post'] }
        }
      };
      delete jsonApiPayload.data.attributes['post'];
    }
    return jsonApiPayload;
  }
```

Is there any way to access registered models in serializer? I guess models `associationKeys` property is best way to determine if key in payload represents a relationship or an attribute isn't it?

**Comment by jelhan** [comment]
If I got it right #880 adds missing serializer method `keyForForeignKey`. This should be enough for the serialize issue. But not quite sure if it also allows us to determine which properties are attributes and which ones relationships in normalize.

**Comment by jrjohnson** [comment]
@jelhan I'm seeing this as well. Did you get around to figuring out what needs to change in the RestSerializer to go from `'user': 1` to `'userId: 1`? If not I will start working on a PR.

**Comment by jelhan** [comment]
@jrjohnson You could use ember-twiddle provided in my first post to get started: https://ember-twiddle.com/328e1448a6a50e8372f0709b4a4be3f6?openFiles=mirage.serializers.application.js%2C

**Comment by samselikoff** [comment]
> I'm also willing to write a PR for ember-cli-mirage RestSerializer to work out of the box with ember-data RESTSerializer. Please let me know if you are interested.

@jelhan absolutely - that would be extremely helpful!

**Comment by tomasznapieralski** [comment]
I had also problem with updating REST models, my fix for that looks like this:

```
// serializers/user.js
import { RestSerializer } from 'ember-cli-mirage';

export default RestSerializer.extend({
  include: ['notifications'],

  normalize() {
    const response = RestSerializer.prototype.normalize.apply(this, arguments);

    response.data.attributes['notifications-id'] = response.data.attributes.notifications;
    delete response.data.attributes.notifications;

    return response;
  },
});
```
I have 1:1 relation between `user` and `notifications` model.

So basically you have to add `-id` suffix to property, which is in a relation.
I know this is hack, but it is the quickest way to fix this.

I hope it will help :)

**Comment by jrjohnson** [comment]
If you want to fix this for all of your relationships you can apply the fix I created in #1121 by creating a `mirage/serializers/application.js` file that contains:

```
import { RestSerializer } from 'ember-cli-mirage';
import { pluralize, camelize, dasherize } from 'ember-cli-mirage/utils/inflector';

export default RestSerializer.extend({
  serializeIds: 'always',
  normalizeIds: 'always',
  normalize(payload) {
    let type = Object.keys(payload)[0];
    let attrs = payload[type];
    let modelName = camelize(type);
    let model = this.schema.modelFor(modelName);
    let { belongsToAssociations, hasManyAssociations } = model.class.prototype;
    let belongsToKeys = Object.keys(belongsToAssociations);
    let hasManyKeys = Object.keys(hasManyAssociations);

    let jsonApiPayload = {
      data: {
        type: pluralize(type),
        attributes: {}
      }
    };
    if (attrs.id) {
      jsonApiPayload.data.id = attrs.id;
    }

    let relationships = {};

    Object.keys(attrs).forEach((key) => {
      if (key !== 'id') {
        if (this.normalizeIds) {
          if (belongsToKeys.includes(key)) {
            let association = belongsToAssociations[key];
            let associationModel = association.modelName;
            relationships[dasherize(key)] = {
              data: {
                type: associationModel,
                id: attrs[key]
              }
            };
          } else if (hasManyKeys.includes(key)) {
            let association = hasManyAssociations[key];
            let associationModel = association.modelName;
            let data = attrs[key].map(id => {
              return {
                type: associationModel,
                id
              };
            });
            relationships[dasherize(key)] = { data };
          } else {
            jsonApiPayload.data.attributes[dasherize(key)] = attrs[key];
          }
        } else {
          jsonApiPayload.data.attributes[dasherize(key)] = attrs[key];
        }
      }
    });
    if (Object.keys(relationships).length) {
      jsonApiPayload.data.relationships = relationships;
    }

    return jsonApiPayload;
  }
});
```

**Comment by erzzo** [comment]
+1
